src/library_pthread.js

```diff
@@ -465,6 +465,16 @@ var LibraryPThread = {
 
     getNewWorker: function() {
       if (PThread.unusedWorkers.length == 0) {
+#if PTHREAD_POOL_SIZE_STRICT || ASSERTIONS
+        var warning = 'Tried to spawn a new thread, but the thread pool is exhausted.\n' +
+        'This will result in a deadlock unless the code explicitly breaks out to the event loop.\n' +
+        'If you want to increase the pool size, use setting `-s PTHREAD_POOL_SIZE=...`.';
+#if !PTHREAD_POOL_SIZE_STRICT
+        err(warning + '\nIf you want to throw an explicit error instead of deadlocking in those cases, use setting `-s PTHREAD_POOL_SIZE_STRICT=1`.');
+#else
+        throw new Error(warning);
```

Comment:
I wonder if we can detect if the new thread was created on the main thread?

When workers themselves create new threads I believe the can be created and started syncronously (because the its the main thread that does the creation).

Perhaps `getNewWorker` -> `getNewWorker(requestComesFromMainThread)`  ?

Reply:
> When workers themselves create new threads I believe the can be created and started syncronously (because the its the main thread that does the creation).

Maybe, but the detection would be tricky, and I see this warning as a stop-gap solution until we implement one of the proper mentioned fixes that would allow to create new threads at-will.

Reply:
This warning is only for folks who use the main thread to start other threads and I think it would be to be precise if we can.  
This will limit the number of folks who end up having to disable this warning because they know what they are doing.

At the very least perhaps we can completely disable this new code if PROXY_TO_PTHREAD is used since in that case the pool is not needed.

Reply:
Also the warning message is wrong for those who start threads from workers.  "This will result in a deadlock unless the code explicitly breaks out to the event loop".. is just not true for workers.. is it?

Reply:
> is just not true for workers.. is it?

Is it always not true or only with PROXY_TO_PTHREAD?

Reply:
Do you perhaps want to add & push those checks to the branch? I have no idea how to implement them and it will be quicker for someone more familiar with codebase.

Reply:
Sure, I've never done collaborative PR before!   I guess you can just undo anything that I push if you don't want to include it.   Or I could do a PR that targets your PR.. lets see.

Reply:
I think it's only with PROXY_TO_PTHREAD btw. E.g. this still hangs even though the second thread is spawned by a Worker:

```cpp
#include <thread>
#include <iostream>

int main()
{
    std::thread([] {
        std::cout << "1\n";
        std::thread([] {
            std::cout << "2\n";
        }).join();
    }).join();
}
```

Reply:
> > is just not true for workers.. is it?
> 
> Is it always not true or only with PROXY_TO_PTHREAD?

I think so... but I could be wrong.  

My understanding is that in PROXY_TO_PTHREAD mode no native user code ever runs on the actual browser thread and that things like new thread creation are all proxied back to the browser thread while user's main thread is free to block as long as it wants.    In this mode the browser thread is supposed to be kind of like an invisible background thread that is more of an implementation detail.

Normally when folks run into the problem this warning is about I say either enable PROXY_TO_PTHREAD *or* add POOL_SIZE=N.. either one will fix the problem.

Reply:
So I guess warning should be hidden if and only if (?) `PROXY_TO_PTHREAD` is used? Maybe under `PROXY_TO_WORKER` too?

Reply:
Added the check to avoid warning/error under `PROXY_TO_PTHREAD`.

Reply:
> I think it's only with PROXY_TO_PTHREAD btw. E.g. this still hangs even though the second thread is spawned by a Worker:
> 
> ```c++
> #include <thread>
> #include <iostream>
> 
> int main()
> {
>     std::thread([] {
>         std::cout << "1\n";
>         std::thread([] {
>             std::cout << "2\n";
>         }).join();
>     }).join();
> }
> ```
> 
> ```shell
> rreverser@rreverser:~/emscripten$ ./emcc temp.cpp -o temp.js -pthread -s PTHREAD_POOL_SIZE=1
> rreverser@rreverser:~/emscripten$ node --experimental-wasm-threads --experimental-wasm-bulk-memory temp.js
> 1
> Tried to spawn a new thread, but the thread pool is exhausted.
> This will result in a deadlock unless the code explicitly breaks out to the event loop.
> If you want to increase the pool size, use setting `-s PTHREAD_POOL_SIZE=...`.
> If you want to throw an explicit error instead of deadlocking in those cases, use setting `-s PTHREAD_POOL_SIZE_STRICT=1`.
> ^C
> ```
> 
> In this case the 2nd warning is still justified and points to a deadlock.



> I think it's only with PROXY_TO_PTHREAD btw. E.g. this still hangs even though the second thread is spawned by a Worker:
> 
> ```c++
> #include <thread>
> #include <iostream>
> 
> int main()
> {
>     std::thread([] {
>         std::cout << "1\n";
>         std::thread([] {
>             std::cout << "2\n";
>         }).join();
>     }).join();
> }
> ```
> 
> ```shell
> rreverser@rreverser:~/emscripten$ ./emcc temp.cpp -o temp.js -pthread -s PTHREAD_POOL_SIZE=1
> rreverser@rreverser:~/emscripten$ node --experimental-wasm-threads --experimental-wasm-bulk-memory temp.js
> 1
> Tried to spawn a new thread, but the thread pool is exhausted.
> This will result in a deadlock unless the code explicitly breaks out to the event loop.
> If you want to increase the pool size, use setting `-s PTHREAD_POOL_SIZE=...`.
> If you want to throw an explicit error instead of deadlocking in those cases, use setting `-s PTHREAD_POOL_SIZE_STRICT=1`.
> ^C
> ```
> 
> In this case the 2nd warning is still justified and points to a deadlock.

Ah I see... in this case the second creation fails because the main thread is blocked so it can't perform the thread creation.

If you remove the outer `join()`  (and move the outer std::thread to global scope) I guess I would work?  (since you are no longer tying up the main thread).

So I guess you are right we can't generally say that worker-originated requests can be serviced ..  :(